src/JustSaying/Fluent/PublishConfig/DynamicMessagePublisher.cs

```diff
@@ -71,4 +72,47 @@ public async Task PublishAsync(Message message, PublishMetadata metadata, Cancel
 
     public Task PublishAsync(Message message, CancellationToken cancellationToken)
         => PublishAsync(message, null, cancellationToken);
+
+    public Task PublishAsync(IEnumerable<Message> messages, CancellationToken cancellationToken)
+        => PublishAsync(messages, null, cancellationToken);
+
+    public async Task PublishAsync(IEnumerable<Message> messages, PublishBatchMetadata metadata, CancellationToken cancellationToken)
+    {
+        var publisherTask = new List<Task>();
+        foreach (var group in messages.GroupBy(x => x.GetType()))
+        {
+            var batchMessages = group.ToList();
+            var message = batchMessages[0];
+
+            var topicName = _topicNameCustomizer(message);
+            if (_batchPublisherCache.TryGetValue(topicName, out var publisher))
+            {
+                publisherTask.Add(publisher.PublishAsync(batchMessages, metadata, cancellationToken));
+                continue;
+            }
+
+            var lockObj = _topicCreationLocks.GetOrAdd(topicName, _ => new SemaphoreSlim(1, 1));
+            _logger.LogDebug("Publisher for topic {TopicName} not found, waiting on creation lock", topicName);
+            await lockObj.WaitAsync(cancellationToken).ConfigureAwait(false);
+            if (_batchPublisherCache.TryGetValue(topicName, out publisher))
+            {
+                _logger.LogDebug("Lock re-entrancy detected, returning existing publisher");
+                publisherTask.Add(publisher.PublishAsync(batchMessages, metadata, cancellationToken));
+                continue;
+            }
+
+            _logger.LogDebug("Lock acquired to initialize topic {TopicName}", topicName);
+            var config = _staticConfigBuilder(topicName);
+            _logger.LogDebug("Executing startup task for topic {TopicName}", topicName);
+            await config.StartupTask(cancellationToken).ConfigureAwait(false);
+
+
+            _ = _batchPublisherCache.TryAdd(topicName, config.BatchPublisher);
```

Comment:
What happens here if `TryAdd()` returns false?

Reply:
Looking at the code, this shouldn't happen since we have a semaphore that allows 1 per time, but let's imagine that happens; I guess we are going to 2 instances of `IMessageBatchPublisher`. GC will collect the newest one after some time.